iosHyperskillApp/iosHyperskillApp/Sources/Modules/StepQuiz/Views/Modals/ProblemsLimitReachedModalViewController.swift

```diff
@@ -0,0 +1,155 @@
+import PanModal
+import SnapKit
+import UIKit
+
+extension ProblemsLimitReachedModalViewController {
+    struct Appearance {
+        let backgroundColor = UIColor.systemBackground
+
+        let contentStackViewSpacing: CGFloat = 32
+        let contentStackViewInsets = LayoutInsets(
+            horizontal: LayoutInsets.defaultInset,
+            vertical: LayoutInsets.largeInset
+        )
+
+        let cubeImageViewSizeRatio = CGSize(width: 1, height: 0.20)
+
+        let goToHomescreenButtonHeight: CGFloat = 44
+    }
+}
+
+final class ProblemsLimitReachedModalViewController: PanModalPresentableViewController {
+    private(set) var appearance = Appearance()
+
+    private lazy var contentStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.spacing = appearance.contentStackViewSpacing
+        stackView.alignment = .leading
+        stackView.distribution = .fill
+        return stackView
+    }()
+
+    override var shortFormHeight: PanModalHeight {
+        let contentStackViewSize = contentStackView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize)
+        let height = appearance.contentStackViewInsets.top + contentStackViewSize.height
+        return .contentHeight(height)
+    }
+
+    override var longFormHeight: PanModalHeight { shortFormHeight }
+
+    private weak var delegate: ProblemsLimitReachedModalViewControllerDelegate?
+
+    init(delegate: ProblemsLimitReachedModalViewControllerDelegate? = nil) {
+        self.delegate = delegate
+        super.init()
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        setup()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        DispatchQueue.main.async {
+            self.panModalSetNeedsLayoutUpdate()
+            self.panModalTransition(to: .shortForm)
+        }
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        delegate?.problemsLimitReachedModalViewControllerDidAppear(self)
+    }
+
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        delegate?.problemsLimitReachedModalViewControllerDidDisappear(self)
+    }
+
+    // MARK: Private API
+
+    private func setup() {
+        view.backgroundColor = appearance.backgroundColor
+
+        setupContentStackView()
+        setupCubeImageView()
+        setupTitleWithTextView()
+        setupGoToHomescreenBackButton()
```

Comment:
You can extract this UI-related code into the UIKit view, for example, `StageImplementUnsupportedModalViewController`. Not necessary to do it right now, but just for your information, it's better to do it like this in the future.